src/cursor/cur_hs.c

```diff
@@ -876,7 +910,7 @@ __curhs_insert(WT_CURSOR *cursor)
 
 retry:
     /* Search the page and insert the updates. */
-    WT_WITH_PAGE_INDEX(session, ret = __wt_hs_row_search(cbt, &file_cursor->key, true));
+    WT_WITH_PAGE_INDEX(session, ret = __curhs_search(cbt, &file_cursor->key, true));
```

Comment:
I think I may find the issue. If we retry search, the key in the file_cursor may have been reset? So we should cache the search key I think.

Reply:
```
/*
     * The pinned page goes away if we do a search, get a local copy of any pinned key or value.
     * Re-save the cursor state: we may retry but eventually fail.
     */
    WT_ERR(__cursor_localkey(cursor));
    WT_ERR(__cursor_localvalue(cursor));
    __cursor_state_save(cursor, &state);
```
@keithbostic What is the purpose of this code in `__bt_cur_insert`?

Reply:
Nice!

In the Btree cursor code, we save a local copy of the key (that is, keys that point into the page with WT_CURSTD_KEY_INT are changed to be WT_CURSTD_KEY_EXT and are copied into the cursor buffer), because once we discard a pinned page in order to do the full tree search, there's a window where a cursor's WT_CURSTD_KEY_INT key can point into a page we no longer have pinned.

I would expect that bug to drop core when the page is evicted, though, I don't see why it would result in a mispositioning of the subsequent operation, unless somehow the key was overwritten in the page, and I don't think that can happen. Even if we're pointing into an insert list instead of the page, the failure mode would be dropping core because we're pointing into random memory when that memory is freed and re-allocated. I mean, *maybe* we'd point into memory that looks like a key, but to fail that way all the time seems unlikely.

I may be missing the path where it fails, though -- what you're describing is definitely a bug and should be fixed, so let's see if it repeats with the appropriate change.

I suspect you can add `WT_ERR(__cursor_localkey(cursor));` right before calling `__wt_cursor_func_init()` in `__curhs_search()` to fix this.